src/transformers/models/llama/modeling_llama.py

```diff
@@ -590,15 +606,129 @@ def _upad_input(self, query_layer, key_layer, value_layer, padding_mask, query_l
+            key_states = torch.cat(key_states, dim=-1)
+
+            value_states = [
+                torch.nn.functional.linear(hidden_states, value_slices[i]) for i in range(self.config.pretraining_tp)
+            ]
+            value_states = torch.cat(value_states, dim=-1)
+        else:
+            query_states = self.q_proj(hidden_states)
+            key_states = self.k_proj(hidden_states)
+            value_states = self.v_proj(hidden_states)
+
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+        value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
+
+        kv_seq_len = key_states.shape[-2]
+        if past_key_value is not None:
+            kv_seq_len += past_key_value[0].shape[-2]
+        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+
+        if past_key_value is not None:
+            # reuse k, v, self_attention
+            key_states = torch.cat([past_key_value[0], key_states], dim=2)
+            value_states = torch.cat([past_key_value[1], value_states], dim=2)
+
+        past_key_value = (key_states, value_states) if use_cache else None
+
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+        # Note that Llama does not use dropout in the attention, hence the hard-coded
+        # dropout_p=0.0 independent of self.training.
+        # The batch_size = 1 case is handled separately to allow to dispatch on flash attention
+        # kernel.
+        if bsz == 1:
+            if query_states.shape[2] > 1:
+                is_causal = True
+            else:
+                is_causal = False
+            attn_output = torch.nn.functional.scaled_dot_product_attention(
+                query_states, key_states, value_states, attn_mask=None, dropout_p=0.0, is_causal=is_causal
+            )
+        else:
+            if attention_mask is not None:
+                if attention_mask.size() != (bsz, 1, q_len, kv_seq_len):
+                    raise ValueError(
+                        f"Attention mask should be of size {(bsz, 1, q_len, kv_seq_len)}, but is {attention_mask.size()}"
+                    )
+
+            attention_mask = torch.max(attention_mask, torch.tensor(torch.finfo(key_states.dtype).min))
+
+            # NOTE: As of PyTorch 2.1, this case can not dispatch to flash attention v2 due to the
+            # attention mask passed. Possible solution: use nested tensors.
+            attn_output = torch.nn.functional.scaled_dot_product_attention(
+                query_states, key_states, value_states, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
+            )
```

Comment:
a lot of this feels duplicated b/w the flash attention 2 implementation and SDP. Could we have a separate method (`_attn`) perhaps that would make it easier to implement various attention optimizations? For example there might be a time where someone might want to use xformers attention implementation.

Reply:
Hi @winglian, yes I agree that a large part of this code is duplicate. Agree as well that having an `_attn` method would make it a bit easier to override the attention (in the same fashion as gpt2, see how it is overwrite for bettertransformer [here](https://github.com/huggingface/optimum/blob/dbe70f9c6675975876e77813b8d0203ac91f075a/optimum/bettertransformer/models/decoder_models.py#L75) and [here](https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/attention.py#L44)).

I wonder whether this is considered breaking change or not though @ArthurZucker @LysandreJik ?

@ydshieh Also suggested to offload the "custom" attention code (flash attention v2, SDPA, etc.) in a separate file for each modeling. Let me know what you think is more elegant and happy to make the change.

ping @patrickvonplaten as well as you have experience with this in diffusers (with the `XFormersAttnProcessor`, `AttnProcessor2_0`, etc.)